spacy/training/initialize.py

```diff
@@ -0,0 +1,205 @@
+from typing import Union, Dict, Optional, Any, List, Callable
+from thinc.api import Config, fix_random_seed, set_gpu_allocator
+from thinc.api import ConfigValidationError
+from pathlib import Path
+import srsly
+
+from .loop import create_before_to_disk_callback
+from ..language import Language
+from ..lookups import Lookups
+from ..errors import Errors
+from ..schemas import ConfigSchemaTraining, ConfigSchemaInit, ConfigSchemaPretrain
+from ..util import registry, load_model_from_config, resolve_dot_names
+from ..util import load_model, ensure_path, logger, OOV_RANK, DEFAULT_OOV_PROB
+
+
+def init_nlp(
+    config: Config,
+    *,
+    use_gpu: int = -1,
+    logger: Callable[[Any], Any] = logger,
+    on_success: Callable[[str], None] = lambda x: None,
+) -> Language:
+    raw_config = config
+    config = raw_config.interpolate()
+    if config["training"]["seed"] is not None:
+        fix_random_seed(config["training"]["seed"])
+    allocator = config["training"]["gpu_allocator"]
+    if use_gpu >= 0 and allocator:
+        set_gpu_allocator(allocator)
+    # Use original config here before it's resolved to functions
+    sourced_components = get_sourced_components(config)
+    nlp = load_model_from_config(raw_config, auto_fill=True)
+    on_success("Set up nlp object from config")
+    config = nlp.config.interpolate()
+    # Resolve all training-relevant sections using the filled nlp config
+    T = registry.resolve(config["training"], schema=ConfigSchemaTraining)
+    dot_names = [T["train_corpus"], T["dev_corpus"]]
+    train_corpus, dev_corpus = resolve_dot_names(config, dot_names)
+    I = registry.resolve(config["initialize"], schema=ConfigSchemaInit)
+    V = I["vocab"]
+    init_vocab(nlp, data=V["data"], lookups=V["lookups"], vectors=V["vectors"])
+    optimizer = T["optimizer"]
+    before_to_disk = create_before_to_disk_callback(T["before_to_disk"])
+    # Components that shouldn't be updated during training
+    frozen_components = T["frozen_components"]
+    # Sourced components that require resume_training
+    resume_components = [p for p in sourced_components if p not in frozen_components]
+    logger.info(f"Pipeline: {nlp.pipe_names}")
+    if resume_components:
+        with nlp.select_pipes(enable=resume_components):
+            logger.info(f"Resuming training for: {resume_components}")
+            nlp.resume_training(sgd=optimizer)
+    with nlp.select_pipes(disable=[*frozen_components, *resume_components]):
+        nlp.begin_training(lambda: train_corpus(nlp), sgd=optimizer)
+        on_success(f"Initialized pipeline components")
+    # Verify the config after calling 'begin_training' to ensure labels
+    # are properly initialized
+    verify_config(nlp)
+    if "pretraining" in config and config["pretraining"]:
+        P = registry.resolve(config["pretraining"], schema=ConfigSchemaPretrain)
+        loaded = add_tok2vec_weights(nlp, P, I)
```

Comment:
The definition of  `add_tok2vec_weights` seems to expect `vocab_config`, which should be `V` instead of `I` in this call, I think?